app/scripts/inpage.js

```diff
@@ -1,56 +1,153 @@
-// mostly a fix for web3's BigNumber if AMD's "define" is defined...
-let __define;
-
-/**
- * Caches reference to global define object and deletes it to
- * avoid conflicts with other global define objects, such as
- * AMD's define function
- */
-const cleanContextForImports = () => {
-  __define = global.define;
-  try {
-    global.define = undefined;
-  } catch (_) {
-    console.warn('MetaMask - global.define could not be deleted.');
-  }
-};
-
-/**
- * Restores global define object from cached reference
- */
-const restoreContextAfterImports = () => {
-  try {
-    global.define = __define;
-  } catch (_) {
-    console.warn('MetaMask - global.define could not be overwritten.');
-  }
-};
-
-cleanContextForImports();
-
 /* eslint-disable import/first */
 import log from 'loglevel';
 import { WindowPostMessageStream } from '@metamask/post-message-stream';
 import { initializeProvider } from '@metamask/providers/dist/initializeInpageProvider';
 
-restoreContextAfterImports();
+(() => {
+  // need to make sure we aren't affected by overlapping namespaces
+  // and that we dont affect the app with our namespace
+  // mostly a fix for web3's BigNumber if AMD's "define" is defined...
+  let __define;
+
+  /**
+   * Caches reference to global define object and deletes it to
+   * avoid conflicts with other global define objects, such as
+   * AMD's define function
+   */
+  const cleanContextForImports = () => {
+    __define = global.define;
+    try {
+      global.define = undefined;
+    } catch (_) {
+      console.warn('MetaMask - global.define could not be deleted.');
+    }
+  };
 
-log.setDefaultLevel(process.env.METAMASK_DEBUG ? 'debug' : 'warn');
+  /**
+   * Restores global define object from cached reference
+   */
+  const restoreContextAfterImports = () => {
+    try {
+      global.define = __define;
+    } catch (_) {
+      console.warn('MetaMask - global.define could not be overwritten.');
+    }
+  };
 
-//
-// setup plugin communication
-//
+  cleanContextForImports();
 
-// setup background connection
-const metamaskStream = new WindowPostMessageStream({
-  name: 'metamask-inpage',
-  target: 'metamask-contentscript',
-});
+  restoreContextAfterImports();
+
+  log.setDefaultLevel(process.env.METAMASK_DEBUG ? 'debug' : 'warn');
+
+  //
+  // setup plugin communication
+  //
+
+  if (shouldInjectProvider()) {
+    // setup background connection
+    const metamaskStream = new WindowPostMessageStream({
+      name: 'metamask-inpage',
+      target: 'metamask-contentscript',
+    });
+
+    initializeProvider({
+      connectionStream: metamaskStream,
+      logger: log,
+      shouldShimWeb3: true,
+    });
+  }
 
-initializeProvider({
-  connectionStream: metamaskStream,
-  logger: log,
-  shouldShimWeb3: true,
-});
+  /**
+   * Determines if the provider should be injected
+   *
+   * @returns {boolean} {@code true} Whether the provider should be injected
+   */
+  function shouldInjectProvider() {
```

Comment:
Could this (or at least some of this) be determined in the worker when injecting inpage.js?   
I think not, because worker couldn't synchronously know some of the inputs here, but worth asking...

Reply:
Nope; the injection happens when the service worker first starts, and there's no callback or logic I can use for a case-by-case basis.